File: backend/app/schemas/environment.py
from typing import Optional, Dict, Any, List
from datetime import datetime
from pydantic import BaseModel, Field


class ConnectionDetails(BaseModel):
    """Connection details for environment systems"""
    fa: Optional[Dict[str, Any]] = Field(None, description="FA environment configuration")
    pya: Optional[Dict[str, Any]] = Field(None, description="PYA environment configuration")
    aws: Optional[Dict[str, Any]] = Field(None, description="AWS environment configuration")
    mbai: Optional[Dict[str, Any]] = Field(None, description="MBAI environment configuration")
    fh: Optional[Dict[str, Any]] = Field(None, description="FH environment configuration")
    system_version: str = Field(..., description="Environment system version")


class EnvironmentStatus(BaseModel):
    """Environment system status response"""
    is_connected: bool = Field(..., description="Environment system connection status")
    environment_system: Optional[str] = Field(None, description="Name of connected environment system")
    iframe_url: Optional[str] = Field(None, description="Authenticated iframe embedding URL")
    external_url: Optional[str] = Field(None, description="External system access URL")
    last_sync: Optional[datetime] = Field(None, description="Last synchronization timestamp")
    connection_details: Optional[ConnectionDetails] = Field(None, description="System connection information")


class EnvironmentLinksResponse(BaseModel):
    """Environment links response"""
    container_id: int = Field(..., description="Container identifier")
    fa: Optional[Dict[str, Any]] = Field(None, description="FA environment configuration")
    pya: Optional[Dict[str, Any]] = Field(None, description="PYA environment configuration")
    aws: Optional[Dict[str, Any]] = Field(None, description="AWS environment configuration")
    mbai: Optional[Dict[str, Any]] = Field(None, description="MBAI environment configuration")
    fh: Optional[Dict[str, Any]] = Field(None, description="FH environment configuration")


class EnvironmentLinksUpdate(BaseModel):
    """Environment links update request"""
    fa: Optional[Dict[str, Any]] = Field(None, description="FA environment configuration")
    pya: Optional[Dict[str, Any]] = Field(None, description="PYA environment configuration")
    aws: Optional[Dict[str, Any]] = Field(None, description="AWS environment configuration")
    mbai: Optional[Dict[str, Any]] = Field(None, description="MBAI environment configuration")
    fh: Optional[Dict[str, Any]] = Field(None, description="FH environment configuration")


class EnvironmentLinksUpdateResponse(BaseModel):
    """Environment links update response"""
    success: bool = Field(..., description="Update success status")
    message: str = Field(..., description="Status message")
    updated_at: datetime = Field(..., description="Update timestamp")


class ContainerContext(BaseModel):
    """Container context for environment system"""
    container_id: int = Field(..., description="Container identifier")
    environment_id: str = Field(..., description="Environment system identifier")


class IframeUrlResponse(BaseModel):
    """FarmHand iframe URL response"""
    iframe_url: str = Field(..., description="Authenticated iframe URL")
    expires_at: datetime = Field(..., description="URL expiration timestamp")
    permissions: List[str] = Field(..., description="User permissions in environment system")
    container_context: ContainerContext = Field(..., description="Container-specific context")


class ExternalUrlResponse(BaseModel):
    """FarmHand external URL response"""
    external_url: str = Field(..., description="Authenticated external URL")
    expires_at: datetime = Field(..., description="URL expiration timestamp")
    session_token: str = Field(..., description="Session authentication token")


class EnvironmentConnectionRequest(BaseModel):
    """Environment system connection request"""
    environment_system: str = Field(..., description="Target environment system")
    fa: Optional[Dict[str, Any]] = Field(None, description="FA environment configuration")
    pya: Optional[Dict[str, Any]] = Field(None, description="PYA environment configuration")
    aws: Optional[Dict[str, Any]] = Field(None, description="AWS environment configuration")
    mbai: Optional[Dict[str, Any]] = Field(None, description="MBAI environment configuration")
    fh: Optional[Dict[str, Any]] = Field(None, description="FH environment configuration")
    user_permissions: List[str] = Field(..., description="Requested user permissions")


class EnvironmentConnectionResponse(BaseModel):
    """Environment system connection response"""
    success: bool = Field(..., description="Connection initiation status")
    message: str = Field(..., description="Status message")
    connection_id: Optional[str] = Field(None, description="Unique connection identifier")
    iframe_url: Optional[str] = Field(None, description="Initial iframe URL")
    external_url: Optional[str] = Field(None, description="Initial external URL")
    estimated_setup_time: Optional[int] = Field(None, description="Estimated setup time in minutes")


class MaintenanceWindow(BaseModel):
    """Scheduled maintenance information"""
    scheduled: bool = Field(..., description="Whether maintenance is scheduled")
    start_time: Optional[datetime] = Field(None, description="Maintenance start time")
    end_time: Optional[datetime] = Field(None, description="Maintenance end time")
    reason: Optional[str] = Field(None, description="Maintenance reason/description")


class EnvironmentSystemHealth(BaseModel):
    """Environment system health status"""
    status: str = Field(..., description="System health status")
    last_heartbeat: datetime = Field(..., description="Last system heartbeat")
    response_time_ms: float = Field(..., description="Average response time in milliseconds")
    system_version: str = Field(..., description="Current system version")
    features_available: List[str] = Field(..., description="Available system features")
    maintenance_window: Optional[MaintenanceWindow] = Field(None, description="Scheduled maintenance information")


class SessionRefreshResponse(BaseModel):
    """Environment system session refresh response"""
    success: bool = Field(..., description="Session refresh status")
    new_iframe_url: Optional[str] = Field(None, description="Updated iframe URL")
    new_external_url: Optional[str] = Field(None, description="Updated external URL")
    expires_at: Optional[datetime] = Field(None, description="New expiration timestamp")
    session_id: Optional[str] = Field(None, description="New session identifier")


class PlaceholderState(BaseModel):
    """Placeholder state for unconnected containers"""
    message: str = Field(default="This container is not yet connected to an environment control system.", 
                        description="Placeholder message")
    can_connect: bool = Field(..., description="Whether user can initiate connection")
    connection_options: Optional[List[str]] = Field(None, description="Available environment systems")

File: frontend/src/types/environment.ts
/**
 * Environment & Recipes Tab Data Models
 * Based on p5_datamodels.md specification
 */

// Connection Details
export interface ConnectionDetails {
  fa?: Record<string, any>;
  pya?: Record<string, any>;
  aws?: Record<string, any>;
  mbai?: Record<string, any>;
  fh?: Record<string, any>;
  system_version: string;
}

// Environment Status
export interface EnvironmentStatus {
  is_connected: boolean;
  environment_system?: string;
  iframe_url?: string;
  external_url?: string;
  last_sync?: string;
  connection_details?: ConnectionDetails;
}

// Environment Links
export interface EnvironmentLinks {
  container_id: number;
  fa?: Record<string, any>;
  pya?: Record<string, any>;
  aws?: Record<string, any>;
  mbai?: Record<string, any>;
  fh?: Record<string, any>;
}

// Container Context
export interface ContainerContext {
  container_id: number;
  environment_id: string;
}

// Iframe Configuration
export interface IframeConfiguration {
  iframe_url: string;
  expires_at: string;
  permissions: string[];
  container_context: ContainerContext;
}

// External URL Configuration
export interface ExternalUrlConfiguration {
  external_url: string;
  expires_at: string;
  session_token: string;
}

// Environment Connection Request
export interface EnvironmentConnectionRequest {
  environment_system: string;
  fa?: Record<string, any>;
  pya?: Record<string, any>;
  aws?: Record<string, any>;
  mbai?: Record<string, any>;
  fh?: Record<string, any>;
  user_permissions: string[];
}

// Environment Connection Response
export interface EnvironmentConnectionResponse {
  success: boolean;
  message: string;
  connection_id?: string;
  iframe_url?: string;
  external_url?: string;
  estimated_setup_time?: number;
}

// Maintenance Window
export interface MaintenanceWindow {
  scheduled: boolean;
  start_time?: string;
  end_time?: string;
  reason?: string;
}

// Environment System Health
export interface EnvironmentSystemHealth {
  status: 'healthy' | 'degraded' | 'down' | 'maintenance';
  last_heartbeat: string;
  response_time_ms: number;
  system_version: string;
  features_available: string[];
  maintenance_window?: MaintenanceWindow;
}

// Session Refresh Response
export interface SessionRefreshResponse {
  success: boolean;
  new_iframe_url?: string;
  new_external_url?: string;
  expires_at?: string;
  session_id?: string;
}

// Placeholder State
export interface PlaceholderState {
  message: string;
  can_connect: boolean;
  connection_options?: string[];
}

// Update Environment Links Request
export interface UpdateEnvironmentLinksRequest {
  fa?: Record<string, any>;
  pya?: Record<string, any>;
  aws?: Record<string, any>;
  mbai?: Record<string, any>;
  fh?: Record<string, any>;
}

// Update Environment Links Response
export interface UpdateEnvironmentLinksResponse {
  success: boolean;
  message: string;
  updated_at: string;
}